### PR TITLE
Mkirk/batch contact intersections

### PR DESCRIPTION
**Pull request by michaelkirk-signal**
PTAL @charlesmchen

### CODE REVIEW

File: SignalServiceKit/src/Messages/OWSMessageSender.m [hunk 1]
```diff
@@ -132,21 +132,9 @@ - (instancetype)initWithMessage:(TSOutgoingMessage *)message
 
 - (nullable NSError *)checkForPreconditionError
 {
-    for (NSOperation *dependency in self.dependencies) {
```
Comment (michaelkirk-signal) [inline SignalServiceKit/src/Messages/OWSMessageSender.m]: I moved this "fail if a dependency failed" logic into the base OWSOperation class so we could use it elsewhere.

File: SignalServiceKit/src/Util/OWSOperation.h [hunk 2]
```diff
@@ -24,22 +24,26 @@ typedef NS_ENUM(NSInteger, OWSOperationState) {
 // any of the errors were fatal. Fatal errors trump retryable errors.
 @interface OWSOperation : NSOperation
 
-@property (nullable) NSError *failingError;
-@property NSUInteger remainingRetries;
+@property (readonly, nullable) NSError *failingError;
```
Comment (michaelkirk-signal) [inline SignalServiceKit/src/Util/OWSOperation.h]: These are all superficial ordering/commenting changes, except I made failingError private-write-only.

File: SignalServiceKit/src/Contacts/OWSContactDiscoveryOperation.swift [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+        Logger.debug("\(logTag) in \(#function) with recipientIdsToLookup: \(recipientIdsToLookup.count)")
+    }
+
+    // MARK: OWSOperationOverrides
+
+    // Called every retry, this is where the bulk of the operation's work should go.
+    override func run() {
+        Logger.debug("\(logTag) in \(#function)")
+
+        Logger.debug("\(logTag) in \(#function) FAKING intersection (TODO)")
```
Comment (michaelkirk-signal) [inline SignalServiceKit/src/Contacts/OWSContactDiscoveryOperation.swift]: stubbed.

File: SignalServiceKit/src/Contacts/OWSContactDiscoveryOperation.swift [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+        guard let cdsOperation = dependencies.first as? CDSBatchOperation else {
+            let error = OWSErrorMakeAssertionError("\(self.logTag) in \(#function) cdsOperation was unexpectedly nil")
+            self.reportError(error)
+            return
+        }
+
+        let cdsRegisteredRecipientIds = cdsOperation.registeredRecipientIds
+
+        if cdsRegisteredRecipientIds == legacyRegisteredRecipientIds {
+            Logger.debug("\(logTag) in \(#function) TODO: PUT /v1/directory/feedback/ok")
```
Comment (michaelkirk-signal) [inline SignalServiceKit/src/Contacts/OWSContactDiscoveryOperation.swift]: stubbed

File: SignalServiceKit/src/Contacts/OWSContactDiscoveryOperation.swift [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+                                        failure: { (task, error) in
+                                            guard let response = task.response as? HTTPURLResponse else {
+                                                let responseError: NSError = OWSErrorMakeUnableToProcessServerResponseError() as NSError
+                                                responseError.isRetryable = true
+                                                self.reportError(responseError)
+                                                return
+                                            }
+
+                                            guard response.statusCode != 413 else {
+                                                let rateLimitError = OWSErrorWithCodeDescription(OWSErrorCode.contactsUpdaterRateLimit, "Contacts Intersection Rate Limit")
```
Comment (charlesmchen-signal) [inline SignalServiceKit/src/Contacts/OWSContactDiscoveryOperation.swift]: FYI this response will eventually include a hint about when to retry.

Comment (charlesmchen-signal) [reply SignalServiceKit/src/Contacts/OWSContactDiscoveryOperation.swift]: Wait, I meant the CDS rate limit errors.

File: SignalServiceKit/src/Contacts/OWSContactDiscoveryOperation.swift [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+
+    // Called every retry, this is where the bulk of the operation's work should go.
+    override func run() {
+        Logger.debug("\(logTag) in \(#function)")
+
+        var phoneNumbersByHashes: [String: String] = [:]
+
+        for recipientId in recipientIdsToLookup {
+            let hash = Cryptography.truncatedSHA1Base64EncodedWithoutPadding(recipientId)
+            phoneNumbersByHashes[hash] = recipientId
```
Comment (charlesmchen-signal) [inline SignalServiceKit/src/Contacts/OWSContactDiscoveryOperation.swift]: assert that there aren't any collisions?

File: SignalServiceKit/src/Contacts/OWSContactDiscoveryOperation.swift [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+
+        Logger.debug("\(logTag) in \(#function) FAKING intersection (TODO)")
+        self.registeredRecipientIds = Set(self.recipientIdsToLookup)
+        self.reportSuccess()
+    }
+}
+
+class CDSFeedbackOperation: OWSOperation {
+
+    static let operationQueue = OperationQueue()
```
Comment (charlesmchen-signal) [inline SignalServiceKit/src/Contacts/OWSContactDiscoveryOperation.swift]: Shouldn't we configure the concurrency level on this queue?

File: SignalServiceKit/src/Contacts/OWSContactDiscoveryOperation.swift [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+    required init(recipientIdsToLookup: [String]) {
+        self.recipientIdsToLookup = recipientIdsToLookup
+        self.registeredRecipientIds = Set()
+
+        super.init()
+
+        Logger.debug("\(logTag) in \(#function) with recipientIdsToLookup: \(recipientIdsToLookup.count)")
+        for batchIds in recipientIdsToLookup.chunked(by: batchSize) {
+            let batchOperation = LegacyContactDiscoveryBatchOperation(recipientIdsToLookup: batchIds)
+            self.addDependency(batchOperation)
```
Comment (charlesmchen-signal) [inline SignalServiceKit/src/Contacts/OWSContactDiscoveryOperation.swift]: So, if any of the operations in the batch fail, we treat the entire batch as a failure, right?  So if any fail, shouldn't we cancel any outstanding operations?

Also, `OWSOperation` does retry, but without exponential backoff.  Don't we need to change that now?

File: SignalServiceKit/src/Contacts/OWSContactDiscoveryOperation.swift (hunk 7 above)
Comment (michaelkirk-signal) [reply SignalServiceKit/src/Contacts/OWSContactDiscoveryOperation.swift]: Good catch! Fixed.

File: SignalServiceKit/src/Contacts/OWSContactDiscoveryOperation.swift (hunk 8 above)
Comment (michaelkirk-signal) [reply SignalServiceKit/src/Contacts/OWSContactDiscoveryOperation.swift]: OWSOperation **can** do retry, but only if you set `remainingRetries` to something greater than 0.

For now, none of these options retry, and I've left all retry logic in the callers (OWSContactManager#intersectContacts:retryDelaySexconds...)

Comment (michaelkirk-signal) [reply SignalServiceKit/src/Contacts/OWSContactDiscoveryOperation.swift]: > So, if any of the operations in the batch fail, we treat the entire batch as a failure, right? So if any fail, shouldn't we cancel any outstanding operations?

Yeah, good idea. I'll make this change.

File: SignalServiceKit/src/Contacts/OWSContactDiscoveryOperation.swift (hunk 6 above)
Comment (michaelkirk-signal) [reply SignalServiceKit/src/Contacts/OWSContactDiscoveryOperation.swift]: done!

File: SignalServiceKit/src/Contacts/OWSContactDiscoveryOperation.swift (hunk 5 above)
Comment (michaelkirk-signal) [reply SignalServiceKit/src/Contacts/OWSContactDiscoveryOperation.swift]: 👍

File: SignalServiceKit/src/Contacts/OWSContactDiscoveryOperation.swift (hunk 8 above)
Comment (michaelkirk-signal) [reply SignalServiceKit/src/Contacts/OWSContactDiscoveryOperation.swift]: I implemented the dependent cancel logic here:

https://github.com/signalapp/Signal-iOS/pull/3726/commits/2e30b33415a3cad161632f9f63f26c467c5f1c87

**Comment by michaelkirk-signal** [comment]
rebased and merged in 0513077d2

**Pull request closed without merging**